### PR TITLE
Typescript migration

### PR DESCRIPTION
**Pull request by tbo47**
Moving to typescript will allow better types and better integration with typescript framework.

@aloisklink @sidharthv96 @remcohaszing Do you mind having a look?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,18 +1,20 @@
...
   "author": "Thibaut Lassalle",
   "license": "MIT",
-  "main": "src/index.js",
+  "main": "dist/index.js",
   "type": "module",
   "scripts": {
-    "generate_types": "tsc --build --clean && tsc --build && prettier --write .",
-    "prepack": "npm run generate_types",
+    "dev": "tsc --watch",
+    "prepublishOnly": "rm -rf dist dist_tmp; mkdir dist; tsc --build tsconfig.json --declaration && mv dist_tmp/src/* dist && rm -rf dist_tmp",
```
Comment (aloisklink) [inline package.json:11]: This should probably be a [`"prepare"` script](https://docs.npmjs.com/cli/v10/using-npm/scripts#life-cycle-scripts) instead so that it runs automatically when somebody does `npm install https://github.com/tbo47/dagre-es`.

File: package.json [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
     "lint": "eslint .",
     "lint:fix": "eslint --fix .",
     "format": "prettier --write .",
     "test": "vitest run",
+    "move_to_typescript": "rename \"s/js$/ts/\" **/*.js",
+    "move_to_javascript": "rename \"s/ts$/js/\" **/*.ts",
```
Comment (aloisklink) [inline package.json:17]: If we no longer need these scripts, can we delete them before merging?

```suggestion
```

File: package.json (hunk 1 above)
Comment (aloisklink) [reply package.json:11]: Also, we could simplify this to just:

```suggestion
    "prepare": "tsc --build tsconfig.json --declaration --clean",
```

if you modify your `tsconfig.json` like:

```json
{
  "include": [
    "src/**/*"
  ],
  "exclude": [
    "src/**/*.test.ts"
  ],
  "compilerOptions": {
    "target": "ES6",
    "rootDir": "src/",
    "outDir": "dist/",
    "lib": ["dom", "dom.iterable", "es2020"],
    "types": ["@types/jest"],
    "module": "node16",
    "skipLibCheck": true
  }
}
```

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -9,20 +9,15 @@ jobs:
...
       - name: Use Node.js ${{ matrix.node-version }}
         uses: actions/setup-node@v3
         with:
           node-version: ${{ matrix.node-version }}
           cache: 'npm'
-      - run: npm ci # throws an error if the package-lock.json file is out of sync
-      - name: Check for linting errors (run `npm run lint:fix` to fix)
-        run: npm run lint
-      - name: Check for prettier errors (run `npm run format` to fix)
-        run: npx prettier --check .
```
Comment (aloisklink) [inline .github/workflows/test.yml:25]: I'm guessing you've disabled this on purpose to make this PR easier to review, so thanks for that!

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -9,20 +9,15 @@ jobs:
...
     steps:
       - uses: actions/checkout@v3
       - name: Use Node.js ${{ matrix.node-version }}
         uses: actions/setup-node@v3
         with:
           node-version: ${{ matrix.node-version }}
           cache: 'npm'
-      - run: npm ci # throws an error if the package-lock.json file is out of sync
-      - name: Check for linting errors (run `npm run lint:fix` to fix)
-        run: npm run lint
```
Comment (aloisklink) [inline .github/workflows/test.yml:23]: Hmmmmm, it will be nice to get https://typescript-eslint.io/ setup, but maybe it's better to do that in a separate PR to make this PR easier to review.

File: package.json [hunk 5]
```diff
@@ -22,25 +24,24 @@
...
-    "typescript": "^4.9.5",
-    "vitest": "^0.29.2"
+    "@types/d3": "^7.4.3",
+    "@types/jest": "^29.5.8",
+    "@types/lodash-es": "^4.17.11",
+    "eslint": "^8.53.0",
+    "eslint-plugin-import": "^2.29.0",
+    "prettier": "^3.1.0",
+    "typescript": "^5.2.2",
+    "vitest": "^0.34.6"
```
Comment (aloisklink) [inline package.json:41]: By the way, GitHub has a free feature called @dependabot that will automatically make PRs updating dependencies if you want: https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuring-dependabot-version-updates#enabling-dependabot-version-updates

It might be worth enabling if you're feeling lazy and want a robot to update everything for you!

**Review by aloisklink — COMMENTED** [review]
Awesome work @tbo47 in doing this! I can also tell that you spent some effort in making this PR as easy to review as possible (e.g. removing autoformatting), so thanks for that.

I've got some comments, but the only major thing I'd really like to see fixed is replacing the `"prepublishOnly"` script in the `package.json` file to a [`"prepare"` script](https://docs.npmjs.com/cli/v10/using-npm/scripts#life-cycle-scripts), so that people can do `npm install https://github.com/tbo47/dagre-es`. All of my other comments are just suggestions.

**Comment by aloisklink** [comment]
Also, we can borrow some types from https://www.npmjs.com/package/@types/dagre, https://www.npmjs.com/package/@types/dagre-d3, and https://www.npmjs.com/package/@types/graphlib if we ever want to improve our types. It's all under the MIT license, so we just need to a copyright message and we can legally use it.

File: package.json (hunk 1 above)
Comment (tbo47) [reply package.json]: Thanks for reviewing!
I tried your suggestion (including the change in tsconfig.json) and it doesn't generate anything in the dist/ folder. So I will update the `prepublishOnly` to `prepare` for now. Don't hesitate to suggest something else.

**Comment by tbo47** [comment]
> Also, we can borrow some types from https://www.npmjs.com/package/@types/dagre, https://www.npmjs.com/package/@types/dagre-d3, and https://www.npmjs.com/package/@types/graphlib if we ever want to improve our types. It's all under the MIT license, so we just need to a copyright message and we can legally use it.

oh nice, I will definitely do that!

**Comment by tbo47** [comment]
> Also, we can borrow some types from https://www.npmjs.com/package/@types/dagre, https://www.npmjs.com/package/@types/dagre-d3, and https://www.npmjs.com/package/@types/graphlib if we ever want to improve our types. It's all under the MIT license, so we just need to a copyright message and we can legally use it.

I just tried to add types but it breaks the unit tests in a way which make me doubt the types in https://www.npmjs.com/package/@types/ are 100% correct. I'm not sure what to do here. Probably work for a later PR...

**Review by sidharthv96 — APPROVED** [review]
LGTM! 🚀

File: src/dagre/position/bk.ts [hunk 6]
```diff
@@ -41,12 +41,12 @@ function findType1Conflicts(g, layering) {
...
 
   function visitLayer(prevLayer, layer) {
-    var // last visited node in the previous layer that is incident on an inner
+    let // last visited node in the previous layer that is incident on an inner
       // segment.
       k0 = 0,
       // Tracks the last node in this layer scanned for crossings with a type-1
       // segment.
-      scanPos = 0,
+      scanPos = 0 as number | string,
```
Comment (remcohaszing) [inline src/dagre/position/bk.ts:49]: `as` is a type cast. It's safer to use a type annotation.

```suggestion
      scanPos: number | string = 0
```

File: src/graphlib/graph.ts [hunk 7]
```diff
@@ -24,7 +24,24 @@ var EDGE_KEY_DELIM = '\x01';
...
+  _defaultEdgeLabelFn: () => any;
+  _nodes: {};
+  _parent: {};
+  _children: {};
+  _in: {};
+  _preds: {};
+  _out: {};
+  _sucs: {};
+  _edgeObjs: {};
+  _edgeLabels: {};
```
Comment (remcohaszing) [inline src/graphlib/graph.ts:41]: These properties are marked public. You can make them private by prepending them with the TypeScript `private` keyword, or you can use real JavaScript private properties

File: tsconfig.json [hunk 8]
```diff
@@ -1,14 +1,11 @@
...
-    "checkJs": true,
-    "declaration": true,
-    "emitDeclarationOnly": true,
-    "lib": ["dom", "dom.iterable", "es2020"],
-    "types": [
-      // vitest types don't work with module "node16", see
-      // https://github.com/vitejs/vite/issues/11552
-      "@types/jest"
-    ],
-    "module": "node16"
```
Comment (remcohaszing) [inline tsconfig.json:12]: You should keep `"module": "node16"`. This is the recommended (only correct really) setting for libraries built with `tsc` (which you do and should)

Comment (aloisklink) [reply tsconfig.json:12]: One thing that I've found really helpful with testing types is https://arethetypeswrong.github.io/ (I think @remcohaszing might have actually been the one that recommended it to me!)

Comment (remcohaszing) [reply tsconfig.json:12]: I might have! It's definitely something I recommend.

But if you just build with `tsc --build` and use `"module": "node16"`, you can't really go wrong.
